gen.ml

```diff
@@ -405,6 +405,20 @@ let take_while p gen =
   take_while (fun x ->x<10) (1--1000) |> eq (1--9)
 *)
 
+let fold_while f s gen =
+  let state = ref s in
+  let gen = take_while (fun e ->
+    let acc, cont = f (!state) e in
```

Comment:
that implementation looks reasonable, but I think it would be better to name the `fun e -> ...`. This way, the use of `take_while` would be clearer:

```ocaml
let folder e =
  let acc, cont = f !state e in
  ...
in
let gen = take_while folder gen in
!state, gen
```

Also, I don't think `fold_while` should return the iterator, only the resulting accumulator. Combinators explicitely discourage using the input generator (see `fold`, `iter`, etc.).